File: binance4j-mining/src/main/java/com/binance4j/mining/client/MiningClient.java
package com.binance4j.mining.client;

import java.util.Map;

import com.binance4j.core.client.RestClient;
import com.binance4j.core.param.FramedPaging;
import com.binance4j.core.param.Paging;
import com.binance4j.core.param.Params;
import com.binance4j.mining.param.AccountListParams;
import com.binance4j.mining.param.AccountProfitsParams;
import com.binance4j.mining.param.AlgorithmsAquisitionParams;
import com.binance4j.mining.param.CoinsAquisitionParams;
import com.binance4j.mining.param.HashrateResaleCancellationParams;
import com.binance4j.mining.param.HashrateResaleDetailParam;
import com.binance4j.mining.param.HashrateResaleListParams;
import com.binance4j.mining.param.HashrateResaleParams;
import com.binance4j.mining.param.MinerDetailsParams;
import com.binance4j.mining.param.MinersParams;
import com.binance4j.mining.param.ProfitsParams;
import com.binance4j.mining.param.StatisticsParams;
import com.binance4j.mining.request.CancelHashrateResaleConfigurationRequest;
import com.binance4j.mining.request.GetAccountProfitstRequest;
import com.binance4j.mining.request.GetAccountsRequest;
import com.binance4j.mining.request.GetAlgorithmsRequest;
import com.binance4j.mining.request.GetCoinsRequest;
import com.binance4j.mining.request.GetHashrateResalesDetailsRequest;
import com.binance4j.mining.request.GetHashrateResalesRequest;
import com.binance4j.mining.request.GetMinersDetailsRequest;
import com.binance4j.mining.request.GetMinersRequest;
import com.binance4j.mining.request.GetOtherProfitsRequest;
import com.binance4j.mining.request.GetProfitsRequest;
import com.binance4j.mining.request.GetStatisticsRequest;
import com.binance4j.mining.request.ResellHashrateRequest;

/**
 * Api client for the NFT endpoints
 * 
 * @see <a href=
 *      "https://binance-docs.github.io/apidocs/spot/en/#mining-endpoints">Documentation</a>
 */
public class MiningClient extends RestClient<MiningMapping> {
	/**
	 * @param key    API public key.
	 * @param secret API secret key.
	 */
	public MiningClient(String key, String secret) {
		super(MiningMapping.class, key, secret);
	}

	/**
	 * Acquiring algorithms.
	 *
	 * @return The request to execute.
	 */
	public GetAlgorithmsRequest getAlgorithms() {
		return new GetAlgorithmsRequest(service.getAlgorithms(new AlgorithmsAquisitionParams().toMap()));
	}

	/**
	 * Acquiring coin name.
	 *
	 * @return The request to execute.
	 */
	public GetCoinsRequest getCoins() {
		return new GetCoinsRequest(service.getCoins(new CoinsAquisitionParams().toMap()));
	}

	/**
	 * Request for detail miner list.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetMinersDetailsRequest getMinersDetails(MinerDetailsParams params) {
		return new GetMinersDetailsRequest(service.getMinersDetails(params.toMap()));
	}

	/**
	 * Request for miner list.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetMinersRequest getMiners(MinersParams params) {
		return new GetMinersRequest(service.getMiners(params.toMap()));
	}

	/**
	 * Get earnings list.
	 *
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetProfitsRequest getProfits(ProfitsParams params, FramedPaging paging) {
		var replace = Map.of("startTime", "startDate", "endTime", "endDate", "page", "pageIndex", "limit", "pageSize");
		return new GetProfitsRequest(service.getProfits(Params.merge(params.toMap(), paging.toMap(replace))));
	}

	/**
	 * Get extra bonus list.
	 *
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetOtherProfitsRequest getOtherProfits(ProfitsParams params, FramedPaging paging) {
		var replace = Map.of("startTime", "startDate", "endTime", "endDate", "page", "pageIndex", "limit", "pageSize");
		return new GetOtherProfitsRequest(service.getOtherProfits(Params.merge(params.toMap(), paging.toMap(replace))));
	}

	/**
	 * Get mining account earning.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAccountProfitstRequest getAccountProfits(AccountProfitsParams params) {
		return new GetAccountProfitstRequest(service.getAccountProfits(params.toMap()));
	}

	/**
	 * Get mining account earning.
	 *
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetAccountProfitstRequest getAccountProfits(AccountProfitsParams params, FramedPaging paging) {
		var replace = Map.of("startTime", "startDate", "endTime", "endDate", "page", "pageIndex", "limit", "pageSize");
		return new GetAccountProfitstRequest(
				service.getAccountProfits(Params.merge(params.toMap(), paging.toMap(replace))));
	}

	/**
	 * Get hashrate resale list.
	 * 
	 * @return The request to execute.
	 */
	public GetHashrateResalesRequest getHashrateResales() {
		return new GetHashrateResalesRequest(service.getHashrateResales(new HashrateResaleListParams().toMap()));
	}

	/**
	 * Get hashrate resale list.
	 *
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetHashrateResalesRequest getHashrateResales(Paging paging) {
		var replace = Map.of("page", "pageIndex", "limit", "pageSize");
		return new GetHashrateResalesRequest(service
				.getHashrateResales(Params.merge(new HashrateResaleListParams().toMap(), paging.toMap(replace))));
	}

	/**
	 * Get hashrate resale detail.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetHashrateResalesDetailsRequest getHashrateResalesDetails(HashrateResaleDetailParam params) {
		return new GetHashrateResalesDetailsRequest(service.getHashrateResalesDetails(params.toMap()));
	}

	/**
	 * Get hashrate resale detail.
	 *
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetHashrateResalesDetailsRequest getHashrateResalesDetails(HashrateResaleDetailParam params,
			Paging paging) {
		var replace = Map.of("page", "pageIndex", "limit", "pageSize");
		return new GetHashrateResalesDetailsRequest(
				service.getHashrateResalesDetails(Params.merge(params.toMap(), paging.toMap(replace))));
	}

	/**
	 * Hashrate resale request.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public ResellHashrateRequest resellHashrate(HashrateResaleParams params) {
		return new ResellHashrateRequest(service.resellHashrate(params.toMap()));
	}

	/**
	 * Cancel hashrate resale configuration.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public CancelHashrateResaleConfigurationRequest cancelHashrateResaleConfiguration(
			HashrateResaleCancellationParams params) {
		return new CancelHashrateResaleConfigurationRequest(service.cancelHashrateResaleConfiguration(params.toMap()));
	}

	/**
	 * Get Statistic list.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetStatisticsRequest getStatistics(StatisticsParams params) {
		return new GetStatisticsRequest(service.getStatistics(params.toMap()));
	}

	/**
	 * Get Account list.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAccountsRequest getAccounts(AccountListParams params) {
		return new GetAccountsRequest(service.getAccounts(params.toMap()));
	}
}

File: src/main/kotlin/com/binance4j/connectors/mining/client/MiningClient.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.binance4j.connectors.mining.client

import com.binance4j.connectors.core.Request
import com.binance4j.connectors.core.client.RestClient

/**
 * Api client for the NFT endpoints
 *
 * [Documentation](https://binance-docs.github.io/apidocs/spot/en/#mining-endpoints)
 */
object MiningClient : RestClient<MiningMapping>(MiningMapping::class.java) {
    /**
     * Acquiring algorithms.
     *
     * @return The request to execute.
     */
    fun getAlgorithms() = Request(service.getAlgorithms())

    /**
     * Acquiring coin name.
     *
     * @return The request to execute.
     */
    fun getCoins() = Request(service.getCoins())

    /**
     * Request for detail miner list.
     *
     * @param algo       Algorithm.
     * @param userName   Mining account.
     * @param workerName Miner’s name.
     *
     * @return The request to execute.
     */
    fun getMinersDetails(algo: String, userName: String, workerName: String) = Request(service.getMinersDetails(algo, userName, workerName))

    /**
     * Request for miner list.
     *
     * @param algo         Algorithm.
     * @param userName     Mining account.
     * @param pageIndex    Page number.
     * @param sort         Sort sequence.
     * @param sortColumn   Miner sort.
     * @param workerStatus Miner status.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getMiners(algo: String, userName: String, workerStatus: String? = null, pageIndex: Int? = null, sort: String? = null, sortColumn: String? = null) =
        Request(service.getMiners(algo, userName, workerStatus, pageIndex, sort, sortColumn))

    /**
     * Get earnings list.
     *
     * @param algo      Transfer algorithm.
     * @param userName  Mining account test.
     * @param coin      Coin name.
     * @param startDate Start time in ms.
     * @param endDate   End time in ms.
     * @param pageIndex Result page.
     * @param pageSize  Results in the page.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getProfits(
        algo: String, userName: String, coin: String? = null,
        startDate: Long? = null, endDate: Long? = null, pageIndex: Int? = null, pageSize: Int? = null
    ) = Request(service.getProfits(algo, userName, coin, startDate, endDate, pageIndex, pageSize))

    /**
     * Get extra bonus list.
     *
     * @param algo      Transfer algorithm.
     * @param userName  Mining account test.
     * @param coin      Coin name.
     * @param startDate Start time in ms.
     * @param endDate   End time in ms.
     * @param pageIndex Result page.
     * @param pageSize  Results in the page.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getOtherProfits(
        algo: String, userName: String, coin: String? = null,
        startDate: Long? = null, endDate: Long? = null, pageIndex: Int? = null, pageSize: Int? = null
    ) = Request(service.getOtherProfits(algo, userName, coin, startDate, endDate, pageIndex, pageSize))

    /**
     * Get mining account earning.
     *
     * @param algo      Transfer algorithm.
     * @param userName  Mining account.
     * @param startDate Start time in ms.
     * @param endDate   End time in ms.
     * @param pageIndex Result page.
     * @param pageSize  Results in the page.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getAccountProfits(
        algo: String, userName: String, startDate: Long? = null, endDate: Long? = null, pageIndex: Int? = null, pageSize: Int? = null
    ) = Request(service.getAccountProfits(algo, userName, startDate, endDate, pageIndex, pageSize))

    /**
     * Get hash rate resale list.
     *
     * @param pageIndex Result page.
     * @param pageSize  Results in the page.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getHashrateResales(pageIndex: Int? = null, pageSize: Int? = null) = Request(service.getHashrateResales(pageIndex, pageSize))

    /**
     * Get hash rate resale detail.
     *
     * @param configId    Config id.
     * @param userName    User name.
     * @param pageIndex   Result page.
     * @param pageSize    Results in the page.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getHashrateResalesDetails(configId: Int, userName: String, pageIndex: Int? = null, pageSize: Int? = null) =
        Request(service.getHashrateResalesDetails(configId, userName, pageIndex, pageSize))

    /**
     * Hash rate resale request.
     *
     * @param userName   Mining Account test.
     * @param algo       Transfer algorithm.
     * @param endDate    Resale End Time in ms.
     * @param startDate  Resale Start Time in ms.
     * @param toPoolUser Mining Account.
     * @param hashRate   Resale hash rate h/s must be transferred (BTC is greater than 500000000000 ETH is greater than 500000).
     *
     * @return The request to execute.
     */
    fun resellHashrate(userName: String, algo: String, startDate: Long, endDate: Long, toPoolUser: String, hashRate: Long) =
        Request(service.resellHashrate(userName, algo, startDate, endDate, toPoolUser, hashRate))

    /**
     * Cancel hash rate resale configuration.
     *
     * @param configId Config id.
     * @param userName User name.
     *
     * @return The request to execute.
     */
    fun cancelHashrateResaleConfiguration(configId: Int, userName: String) =
        Request(service.cancelHashrateResaleConfiguration(configId, userName))

    /**
     * Get Statistic list.
     *
     * @param algo     Algorithm.
     * @param userName Mining account.
     *
     * @return The request to execute.
     */
    fun getStatistics(algo: String, userName: String) = Request(service.getStatistics(algo, userName))

    /**
     * Get Account list.
     *
     * @param algo     Algorithm.
     * @param userName Mining account.
     *
     * @return The request to execute.
     */
    fun getAccounts(algo: String, userName: String) = Request(service.getAccounts(algo, userName))
}